content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,643 @@
+  Thus, extra care is needed when using `ref`s within functions like [`createMemo`](/reference/createMemo), [`createRenderEffect`](/reference/createRenderEffect), and [`createComputed`](/reference/createComputed).
+</Aside>
+
+Lastly, there's a less-safe pattern where the definite assignment assertion is declared right when the variable is initialized.
+This technique bypasses TypeScript's assignment checks for that specific variable, providing a quick workaround at the cost of potential runtime errors.
+
+```typescript
+let divRef!: HTMLDivElement;
+// This will be permitted but will throw an error at runtime
+divRef.focus();
+onMount(() => {
+  divRef.focus(); 
+});
+return <div ref={divRef}>...</div>;
+```
+
+## Control flow narrowing
+
+When using control flow components in Solid, such as [`<Show>`](/reference/show), it's common to conditionally display content based on the presence of certain data.
+This, however, can introduce some type-checking challenges, especially when attempting to manipulate the daata within these components.
+
+Consider a situation where you want to show a name only when it's defined using a Solid signal:
+
+```typescript
+const [name, setName] = createSignal<string>();
+return <Show when={name()}>Hello {name().replace(/\s+/g, "\xa0")}!</Show>;
+```
+
+TypeScript may not like this.
+While you can understand that the second call to `name()` happens only if the first is truthy, TypeScript cannot. 
+It may flag an error indicating that `name()` could potentially be `undefined` when `.replace()` is called.
+
+There are a few ways to tackle this:
+
+1. **Non-Null Assertion**
+Use TypeScript's [non-null assertion operator (!).](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) 
+This operator tells TypeScript you're confident the value isn't `null` or `undefined`:
+
+```typescript
+return <Show when={name()}>Hello {name()!.replace(/\s+/g, "\xa0")}!</Show>;
+```
+While this may be a quick option, incorrectly asserting can lead to runtime errors if the value does turn out to be `null` or `undefined`.
+
+2. **Callback form of Show**
+
+Solid offers a callback form for the [`<Show>`](/reference/show) component, where the truthy value of the `when` prop gets passed in:
+
+```typescript
+const [data, setData] = createSignal<User | Admin>() // T: Signal<User | Admin | undefined>
+
+return(
+    <Show when={() => {
+      const v = data()
+      return v && v.type === "user" && v // T: User | undefined
+    })}>
+      {user => <User user={user()} />}
+    </Show>
+  )
+```
+
+This method leverages the type information embedded in the [`<Show>`](/reference/show) component.
+TypeScript can infer that since `n` comes from a truthy `name()`, it can't be `undefined` or other falsy values.
+However, accessing this null asserted form incorrectly may thrown an error if the condition is not satisfied.
```

Comment:
```suggestion
However, accessing the callback may throw an error if the `when` prop is nullish or false at the time of access. Usually this is not the case, however in `onCleanup` or asynchronous contexts it can be.
```
In general you should never attempt to read signals in `onCleanup` or async, but always store their value, unless you specifically want their latest value; Most of the time in `onCleanup` you want the value when `onCleanup` was called.